text/0000-module-resolution.md

```diff
@@ -0,0 +1,356 @@
+
+::: note
+  The folder structure is something that the core LWC teams wants to enforce as a convention. There is no technical limitation on the compiler or other tools to resolve different folder structures. However we do want to provide and enforce as much as possible a canonical way so we can in the future keep iterating, improving and expanding the ergonomics of it.
+:::
+
+
+```json
+{
+    "modules": [
+       { "dir" : "src/modules" }
+    ]
+}
+```
+
+Here is an example of a valid module resolution given the previous DirectoryModuleRecord
+```
+src
+└── modules/ 
+    └── ui/ (namespace)
+        └── button/ (moduleName)
+            └── button.js (moduleEntry)
+```
+
+`DirectoryModuleRecord` must contain `dir` key.
+
+#### NpmModuleRecord
+
+This type of resolution tells the resolver to find an npm package with that name and resolve the modules it might have inside. This process can be recursive (see example at the top).
+
+```json
+{
+    "modules": [
+       { "npm" : "@ui/components" }     
+    ]
+}
+```
+
+`NpmModuleRecord` must contain `npm` key.
+
+## Algorithm
+
+Here is the algoritm for module resolution (the instructions are not in a rigurous specification format for brevity and time)
+
+```
+  Let `rootDir` be the initial path for the application, `initModules` 
+  the list of ModuleRecords to be resolved and ModuleRecordEntryList 
+  an empty array.
+  
+  1. Load the LWC configuration file in the rootDir
+    1.1 Search for `lwc.config.js`, if it exist read it.
+    1.2 If lwc.config.js is not found search in package.json
+    (If both are found the configuration file has precedence).
+
+  2. Merge all the ModuleRecords found in the config with 
+  the list of initModules. If an alias with the same name 
+  exist pick initModules over the ones resolved in the config.
+
+  3. For each ModuleRecord:
+
+    3.1  Validate that its a valid ModuleRecord type, skip it otherwise.
+    3.1.1 For `AliasModuleRecord` `path` and `name` keys must exist.
+    3.1.2 For `DirModuleRecord` the dir key must exist and the 
+    value must match an existing folder on the file system.
+    3.1.3 For `NpmModuleRecord` the npm key must exist and its value 
+    must not have a leading `/` or `./`. Also the npm package must
+    be resolvable from Node's perspective.
+
+    3.2 Resolve ModuleRecord entry:
+
+        3.2.1 If is an `AliasModuleRecord` validate the path and add
+        create a ModuleRecordEntry with:
+          - `specifier` as the `name` value.
+          - `scope` with the closest configuration path.
+          - `entry` with the `path` value.
+
+        3.2.2 If is a `DirModuleRecord` validate the path and find of 
+        the modules that match the structure:
+        [namespace]/[componentName]/[componentName.{html|css|js|ts}]
```

Comment:
```suggestion
        [namespace]/[componentName]/[componentName.{html|css|js|ts}] and create a 
        ModuleRecordEntry with:
          - `specifier` be `[namespace]/[componentName]`
          - `scope` be the closest configuration path
          - `entry` be `[dir]/[namespace]/[componentName]/[componentName.{html|css|js|ts}]`
```

This section of the spec is missing in which order the extension should be resolved. For example if a module contains `test.html`, `test.css`, `test.js` and `test.ts` which file is considered the entry point. I am proposing the following order:
- `.js`
- `.ts`
- `.html`
- `.css`